src/main/java/org/opensearch/security/configuration/SecurityFlsDlsIndexSearcherWrapper.java

```diff
@@ -87,78 +94,113 @@ public SecurityFlsDlsIndexSearcherWrapper(
         this.clusterService = clusterService;
         this.indexService = indexService;
         this.auditlog = auditlog;
-        this.dlsQueryParser = new DlsQueryParser(indexService.xContentRegistry());
         final boolean allowNowinDlsQueries = settings.getAsBoolean(ConfigConstants.SECURITY_UNSUPPORTED_ALLOW_NOW_IN_DLS, false);
         if (allowNowinDlsQueries) {
             nowInMillis = () -> System.currentTimeMillis();
         } else {
             nowInMillis = () -> { throw new IllegalArgumentException("'now' is not allowed in DLS queries"); };
         }
         log.debug("FLS/DLS {} enabled for index {}", this, indexService.index().getName());
-        this.salt = salt;
+        this.dlsFlsProcessedConfigSupplier = dlsFlsProcessedConfigSupplier;
+        this.dlsFlsBaseContext = dlsFlsBaseContext;
     }
 
     @SuppressWarnings("unchecked")
     @Override
     protected DirectoryReader dlsFlsWrap(final DirectoryReader reader, boolean isAdmin) throws IOException {
 
         final ShardId shardId = ShardUtils.extractShardId(reader);
+        PrivilegesEvaluationContext privilegesEvaluationContext = this.dlsFlsBaseContext.getPrivilegesEvaluationContext();
 
-        Set<String> flsFields = null;
-        Set<String> maskedFields = null;
-        Query dlsQuery = null;
-
-        if (!isAdmin) {
+        if (log.isTraceEnabled()) {
+            log.trace("dlsFlsWrap(); index: {}; privilegeEvaluationContext: {}", index.getName(), privilegesEvaluationContext);
+        }
 
-            final Map<String, Set<String>> allowedFlsFields = (Map<String, Set<String>>) HeaderHelper.deserializeSafeFromHeader(
-                threadContext,
-                ConfigConstants.OPENDISTRO_SECURITY_FLS_FIELDS_HEADER
-            );
-            final Map<String, Set<String>> queries = (Map<String, Set<String>>) HeaderHelper.deserializeSafeFromHeader(
-                threadContext,
-                ConfigConstants.OPENDISTRO_SECURITY_DLS_QUERY_HEADER
-            );
-            final Map<String, Set<String>> maskedFieldsMap = (Map<String, Set<String>>) HeaderHelper.deserializeSafeFromHeader(
+        if (isAdmin || privilegesEvaluationContext == null) {
```

Comment:
The `PrivilegeEvaluationContext` follows the same "nullness" semantics as `User`: If this is a system internal request without an initiating user, this will be `null`.